### PR TITLE
[#407] CodeView: give additional sorting options

### LINKED ISSUE #407: code view: give sorting options
Current: auto-sorted by LoC attributed to the author

Suggested: give an option to sort by other options e.g., file name, file type

### PR DESCRIPTION
**Pull request by emer7**
Fix #407 
```
Files in the code view are automatically sorted by LoC.

While this helps reviewers to quickly locate the authors' top
contributed files, some may prefer looking through the files in other
hierarchical order.

Let's add sorting selection in the code view, allow users to
sort the files in different order (file name, file type or file path),
catering to the need of diversified users. 

```

### CODE REVIEW

**Comment by emer7** [comment]
Ready for comments

**Comment by emer7** [comment]
> Perhaps making the sorting option as a `select` menu like in the chart panel would be better (we would know how it is currently sorted)? And also having a checkbox to reverse the sorting order as well.

Sure okay. I will implement it to follow the main page to be consistent in style.

> Can also add an option to sort by file types as mentioned in #407 

Okay I thought Chelsey has implemented that

**Comment by emer7** [comment]
Implemented select menu with reverse checkbox and sort by file type.

**Comment by emer7** [comment]
Ready for review again

**Comment by eugenepeh** [comment]
> how to repro it?

I am using the `developer tools` > `console` of the chrome browser.

**Comment by emer7** [comment]
> > how to repro it?
> 
> I am using the `developer tools` > `console` of the chrome browser.

I don't get it in my machine :/
do you do something specific?

**Comment by eugenepeh** [comment]
>I don't get it in my machine :/
do you do something specific?

Need to keep it open and use the feature that you've implemented. 

image

**Comment by eugenepeh** [comment]
>It seems the violation come from , and it also plagues in the summary pane. It appears in master branch

Hmm okay, opened issue #517. Any idea what causes it?

**Comment by eugenepeh** [comment]
Nits to the proposed message:

~RepoSense have a code panel where user can view the files written
by an author in a repo.~ `(Can omit logical details and go straight to the point)`ln the panel the files are automatically sorted by LoC.
User ~would want~ `(Try to use phases like may want etc instead, because there are no backing evidences to your words)` to ~sort by other means such as file name, file type, etc.~ `(Avoid duplicating the reason for cause)`

Let's implement sorting by file name, file type, and path in the code panel.

Modified:
```
Files in the code view are automatically sorted by LoC.

While this helps reviewers to quickly locate the authors' top
contributed files, some may prefer looking through the files in other
hierarchical order.

Let's add sorting selection in the code view, allow users to
sort the files in different order (file name, file type or file path),
catering to the need of diversified users.

**Comment by eugenepeh** [comment]
>Doesnt' seem to work at some situations:

Yes... seems like it stop working once any of the file types are de-selected.

**Comment by emer7** [comment]
working now

**Comment by yamidark** [comment]
@emer7 There is now ESLint checks in Travis, do resolve them as well.

**Comment by emer7** [comment]
@reposense/devs ready for review again

**Comment by emer7** [comment]
> Doesn't seems to be handling the background work.
> 
> image
> 
> When the sorting selection is changed when any of the file formats are deselected, sorting were not performed along with the "hidden" files which caused it to show up in jumbled order when its selected to be shown again.
> 
> On a side note, do change the `File Types` to `File Formats` to make it standardize across all documentations.

@eugenepeh fixed

**Comment by eugenepeh** [comment]
Still doesn't seems to be working.

image

**Comment by yamidark** [comment]
@emer7 Is this ready for review? Do tag the other @reposense/devs people to review your work at right side of this page when it is.

**Comment by emer7** [comment]
Yes it is ready to be reviewed again. I cannot ask @eugenepeh to review again though.

File: frontend/src/static/js/v_authorship.js [hunk 1]
```diff
@@ -165,12 +184,21 @@ window.vAuthorship = {
           }), {});
     },
 
+    filesSort() {
```
Comment (eugenepeh) [inline frontend/src/static/js/v_authorship.js]: naming wise, should follow existing style. 
action follow by nouns.
E.g. sortFiles,

File: frontend/src/static/js/v_authorship.js [hunk 2]
```diff
@@ -165,12 +184,21 @@ window.vAuthorship = {
...
+      this.selectedFiles.sort(window.comparator(filesSortDict[this.filesSortType]));
+
+      if (this.filesSortReverse) {
```
Comment (eugenepeh) [inline frontend/src/static/js/v_authorship.js]: same as the above.
name it as reverseSortFiles. 
Also, to comply with our coding convention, the boolean variables/methods should be named to sound like booleans.
E.g. prefixed with `is`, `can`, `should`, `to` etc. 
For this case, let's name it `toReverseSortFiles`.

File: .project [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+		<buildCommand>
+			<name>org.eclipse.buildship.core.gradleprojectbuilder</name>
+			<arguments>
+			</arguments>
+		</buildCommand>
+	</buildSpec>
+	<natures>
+		<nature>org.eclipse.buildship.core.gradleprojectnature</nature>
+	</natures>
+</projectDescription>
```
Comment (yamidark) [inline .project]: Ehh, sorry if I missed anything, but may I know what is this file for?
If i'm not wrong its the Eclipse IDE project description file right? This shouldn't be inside our repo 😱 .

Comment (emer7) [reply .project]: Apology for the oversight. Must have accidentally commited it when activating Java debugging

**Comment by emer7** [comment]
> Seems like some error is thrown when trying to open the codeview (sometimes), causing it to crash:
> untitled

I cannot repro it. How do you get this error? @yamidark

**Comment by yamidark** [comment]
>I cannot repro it. How do you get this error? @yamidark 

I used netlify to check this back then. Seems to be resolved now though.
Do help to resolve the conflicts and request for review again

**Comment by damithc** [comment]
The `reverse` option doesn't seem to work correctly. In the example below, ticking the option changes the sort order but unticking it doesn't change the sort order.
image

**Comment by emer7** [comment]
@reposense/stage2-reviewers ready for review again

**Comment by emer7** [comment]
> The `reverse` option doesn't seem to work correctly. In the example below, ticking the option changes the sort order but unticking it doesn't change the sort order.
> image

Fixed Prof @damithc

**Comment by emer7** [comment]
@eugenepeh fixed

**Comment by emer7** [comment]
> 👍
> 
> Just a few more aesthetic tweaks:
> 
> image
> 
> * Let's move the sorting to the top, leaving the radio controls at the bottom to emphasize the number of radio buttons better.
> * The gap between the sorting and radio controls need to be wider, preferably same distance like the 2 radio button divs.
> * The font sizes of 3 controls are different (search bar, checkbox, sort); should make it the same.
> * The font reverse checkbox should be black to differentiate it from other checkboxes and to be in consistent with chart view.

@eugenepeh is this being fixed by @fzdy1914 in PR #589 ?

**Comment by yamidark** [comment]
> @eugenepeh is this being fixed by @fzdy1914 in PR #589 ?

Oops, I thought this was done already 😓. Need to double-check again in the future before I merge....

Anyway @fzdy1914 do you mind helping to do this fixes on your PR #589 as well (shouldn't be too much)? Opening another PR for this would most probably result in more conflicts to resolve again.

**Comment by fzdy1914** [comment]
> > @eugenepeh is this being fixed by @fzdy1914 in PR #589 ?
> 
> Oops, I thought this was done already 😓. Need to double-check again in the future before I merge....
> 
> Anyway @fzdy1914 do you mind helping to do this fixes on your PR #589 as well (shouldn't be too much)? Opening another PR for this would most probably result in more conflicts to resolve again.

OK, I will fix this together in #589

**Comment by eugenepeh** [comment]
>@eugenepeh is this being fixed by @fzdy1914 in PR #589 ?

Technically it shouldn't because many of the change has no relevance to the checkbox alignment.
Do open another PR as the make up fix

**Pull request merged**